### ISSUE TITLE
New single file bundle loading slow due to AV intervention

### ISSUE THREAD

**Issue by dasMulli** [dotnet/runtime#47621]
We updated some tools and services to .NET 5 and in the process single file deployments changed from extraction to bundle-loading (not using native dependency or content file extraction).

While not so much a problem for services, for command line tools the startup time worsened dramatically to due to AV scanning (McAfee with on-access scanning).
Previously this wasn't a problem because files extracted by the single file bundle host into the temp directory were scanned once by AV and then apparently marked safe and cached.

However the .NET 5 bundle seems to trigger scanning every time it is launched.
Is this a known behavior with enterprise AV? Does this show up in MS' AV simulator? (That helped diagnosing extraction issues on previous issues I created).

Is there a way to use the previous bundling logic in .NET 5 where all managed DLLs are extracted to a temporary location?

Consider a sample self-contained single file app referencing that is run on Windows Server 2019 (Dual Xeon 6248, 512GB RAM) with enterprise McAfee installed.:

.NET 3.1 (after repeated execution):
```
> Measure-Command { .\our-util.exe --help }
...
Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 122
Ticks             : 1226997
TotalDays         : 1,42013541666667E-06
TotalHours        : 3,408325E-05
TotalMinutes      : 0,002044995
TotalSeconds      : 0,1226997
TotalMilliseconds : 122,6997
```

.NET 5 (after repeated execution):
```
> Measure-Command { .\our-util.exe --help }
...
Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 3
Milliseconds      : 242
Ticks             : 32424450
TotalDays         : 3,75282986111111E-05
TotalHours        : 0,000900679166666667
TotalMinutes      : 0,05404075
TotalSeconds      : 3,242445
TotalMilliseconds : 3242,445
```
During the additional 3 seconds we can see McAfee CPU usage spiking.

The .NET 5 variant without single-file (but still self-contained) shows `TotalMilliseconds : 110,3066`

While for us with the mentioned program the issue is only a bit annoying at best, I fear that this raises larger concerns for shipping self-contained command line tools or utility applications in enterprise scenarios.

**Comment by ThomasGoulet73** [comment]
If you want the behavior of .Net Core 3, you could use `IncludeAllContentForSelfExtract`. https://docs.microsoft.com/en-us/dotnet/core/deploying/single-file#other-considerations

**Comment by Fabi** [comment]
That doesn't sound like a .NET 5 related problem though. Files that are extracted are still in the same temp folder and the executing file is usually only scanned once unless specified otherwise.

**Comment by agocke** [comment]
In .NET 5, by default we no longer extract the assemblies, we just load them directly from memory. It may be the case that the AV sees the module load, but doesn't know how to cache the load results. Unfortunately I don't really have an answer for that -- it sounds like the AV provider needs to improve performance.

**Comment by vitek-karas** [comment]
The runtime calls into the [AMSI](https://docs.microsoft.com/en-us/windows/win32/amsi/antimalware-scan-interface-portal) to report any loads from memory. The caching of results for this will be tricky. We will probably have to talk to security people to figure out if there's something we can do about this.

**Comment by Fabi** [comment]
> The runtime calls into the [AMSI](https://docs.microsoft.com/en-us/windows/win32/amsi/antimalware-scan-interface-portal) to report any loads from memory. The caching of results for this will be tricky. We will probably have to talk to security people to figure out if there's something we can do about this.

I'm sure this already got some kind of caching for similar memory content that was used before. Otherwise apps would be pretty slow in general which isn't the case.

**Comment by vitek-karas** [comment]
@Fabi - I'm not sure about that - the runtime only uses AMSI for assemblies loaded from memory. Those loaded from disk are intercepted by the OS. And I assume OS caches the results based on the timestamps of the files on disk - so it's cheap to check the validity of the cache. For memory loaded assemblies, this might be a bit more tricky (at the very least the entire assembly would have to be read and something like a hash of the content used to determine if a result was already calculated for it).

In short - most of "normal" apps will almost exclusively rely on the file based AV scanning, and that is probably faster because of the above reasoning. But I'm really just guessing - we would need to talk to security experts and AV people about this.

**Comment by vitek-karas** [comment]
Just to set the expectations on this (I quickly chatted with our security team about this): There's no way we will omit calling AMSI for code loaded from the single-file executable. I'm not aware of anything we can do here to "fix" this really.

**Comment by agocke** [comment]
@vitek-karas is right -- there's not much in our control here. The AV vendor can cache assemblies they've already seen, but we don't have any way to force or recommend that.
